src/data/ar-io-data-source.ts

```diff
@@ -162,22 +171,61 @@ export class ArIODataSource implements ContiguousDataSource {
       throw new Error('No weighted peers available');
     }
 
-    return randomWeightedChoices<string>({
-      table: this.weightedPeers,
-      count: peerCount,
-    });
+    return shuffleArray([
+      ...this.getRandomWeightedPeers(this.weightedPeers, peerCount),
+    ]);
   }
 
-  handlePeerSuccess(peer: string): void {
+  handlePeerSuccess(peer: string, kbps: number, ttfb: number): void {
     metrics.getDataStreamSuccessesTotal.inc({
       class: this.constructor.name,
       source: peer,
     });
+
+    this.previousGatewayPeerTtfbDurations.push(ttfb);
+    if (
+      this.previousGatewayPeerTtfbDurations.length >
+      config.GATEWAY_PEERS_REQUEST_WINDOW_COUNT
+    ) {
+      this.previousGatewayPeerTtfbDurations.shift();
+    }
+
+    this.previousGatewayPeerKbpsDownloadRate.push(kbps);
+    if (
+      this.previousGatewayPeerKbpsDownloadRate.length >
+      config.GATEWAY_PEERS_REQUEST_WINDOW_COUNT
+    ) {
+      this.previousGatewayPeerKbpsDownloadRate.shift();
+    }
+    const currentAvarageTtfb =
+      this.previousGatewayPeerTtfbDurations.length === 0
+        ? 0
+        : this.previousGatewayPeerTtfbDurations.reduce(
+            (acc, value) => acc + value,
+            0,
+          ) / this.previousGatewayPeerTtfbDurations.length;
+
+    const currentAverageKbps =
+      this.previousGatewayPeerKbpsDownloadRate.length === 0
+        ? 0
+        : this.previousGatewayPeerKbpsDownloadRate.reduce(
+            (acc, value) => acc + value,
+            0,
+          ) / this.previousGatewayPeerKbpsDownloadRate.length;
+
+    const additionalWeightFromTtfb =
+      ttfb >= currentAvarageTtfb ? config.WEIGHTED_PEERS_TEMPERATURE_DELTA : 0;
```

Comment:
I think we may ultimately want different weight deltas for these, but not need to complicate things now.

Reply:
I think this comparison is backwards. Low `ttfb` is good.